### PR TITLE
Implement callback support of async_send_request for service generic client

### PR DESCRIPTION
**Pull request by Barry-Xu-2018**
Currently, the implementation of `async_send_request` for the GenericClient lacks support for callbacks.   
When a response is received, the user-defined callback can be automatically invoked.

### CODE REVIEW

File: rclcpp/test/rclcpp/test_generic_client.cpp [hunk 1]
```diff
@@ -228,3 +229,69 @@ TEST_F(TestGenericClientSub, construction_and_destruction) {
     }, rclcpp::exceptions::InvalidServiceNameError);
   }
 }
+
+TEST_F(TestGenericClientSub, async_send_request_with_request) {
+  const std::string service_name = "test_service";
+  int64_t expected_change = 1111;
```
Comment (ahcorde) [inline rclcpp/test/rclcpp/test_generic_client.cpp:235]: include `<cstdint>`

File: rclcpp/test/rclcpp/test_generic_client.cpp [hunk 2]
```diff
@@ -228,3 +229,69 @@ TEST_F(TestGenericClientSub, construction_and_destruction) {
...
+  auto client = node->create_generic_client(service_name, "test_msgs/srv/BasicTypes");
+
+  auto callback = [&expected_change](
+    const test_msgs::srv::BasicTypes::Request::SharedPtr request,
+    test_msgs::srv::BasicTypes::Response::SharedPtr response) {
+      response->int64_value = request->int64_value + expected_change;
+    };
+
+  auto service =
+    node->create_service<test_msgs::srv::BasicTypes>(service_name, std::move(callback));
```
Comment (ahcorde) [inline rclcpp/test/rclcpp/test_generic_client.cpp:246]: include `<utility>`

File: rclcpp/test/rclcpp/test_generic_client.cpp [hunk 3]
```diff
@@ -228,3 +229,69 @@ TEST_F(TestGenericClientSub, construction_and_destruction) {
...
+
+  ASSERT_TRUE(client->wait_for_service(std::chrono::seconds(5)));
```
Comment (ahcorde) [inline rclcpp/test/rclcpp/test_generic_client.cpp:248]: include `<chrono>`

**Review by ahcorde — CHANGES_REQUESTED** [review]

File: rclcpp/test/rclcpp/test_generic_client.cpp (hunk 1 above)
Comment (Barry-Xu-2018) [reply rclcpp/test/rclcpp/test_generic_client.cpp:235]: Thanks, I will add it.

File: rclcpp/test/rclcpp/test_generic_client.cpp (hunk 2 above)
Comment (Barry-Xu-2018) [reply rclcpp/test/rclcpp/test_generic_client.cpp:246]: This header file already exists.

File: rclcpp/test/rclcpp/test_generic_client.cpp (hunk 3 above)
Comment (Barry-Xu-2018) [reply rclcpp/test/rclcpp/test_generic_client.cpp:248]: I will add it

**Comment by sloretz** [comment]
@fujitatomoya would you be willing to give this one another review? 🧇

File: rclcpp/test/rclcpp/test_generic_client.cpp [hunk 4]
```diff
@@ -228,3 +231,69 @@ TEST_F(TestGenericClientSub, construction_and_destruction) {
     }, rclcpp::exceptions::InvalidServiceNameError);
   }
 }
+
+TEST_F(TestGenericClientSub, async_send_request_with_request) {
```
[the remark this answers is not in the document]
Comment (Barry-Xu-2018) [reply rclcpp/test/rclcpp/test_generic_client.cpp:235]: I also considered putting the test in test_client_common.cpp.

In test_client_common.cpp, all tests are based on message "test_msgs/srv/Empty".  A test can be applicable to 2 different types of clients simultaneously.

For the tests in this PR, they need to the message type "test_msgs/srv/BasicTypes". Even if the test moves to test_client_common.cpp, I only want to only test GenericClient.  

So I ended up putting the test in test_generic_client.cpp.

[the remark this answers is not in the document]
Comment (Barry-Xu-2018) [reply rclcpp/test/rclcpp/test_generic_client.cpp:235]: @fujitatomoya Please take a look at my above comments.

**Review by fujitatomoya — APPROVED** [review]
lgtm with green CI

**Comment by fujitatomoya** [comment]
@ahcorde could you have a 2nd review?

**Comment by fujitatomoya** [comment]
Pulls: ros2/rclcpp#2614
Gist: https://gist.githubusercontent.com/fujitatomoya/6b080ca006868c0a0419a2a612cd38ae/raw/e03f6ad7967881eacf9d8bc611917e0de0f0701d/ros2.repos
BUILD args:  --packages-above-and-dependencies rclcpp
TEST args:  --packages-above rclcpp
ROS Distro: rolling
Job: ci_launcher
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/14577
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/21854/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/16124/)
* Linux-rhel [Build Status](http://ci.ros2.org/job/ci_linux-rhel/1360/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/22565/)

**Comment by Barry-Xu-2018** [comment]
Missing this information
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/22565/)  

I will fix this window issue.

**Comment by Barry-Xu-2018** [comment]
Re-run CI on Windows

* Windows [Build Status](http://ci.ros2.org/job/ci_windows/22625/)

**Comment by ahcorde** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/21907/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/16175/)
* Linux-rhel [Build Status](http://ci.ros2.org/job/ci_linux-rhel/1416/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/22628/)

**Review by ahcorde — APPROVED** [review]

**Pull request merged**